src/docfx/build/xref/XrefResolver.cs

```diff
@@ -93,35 +93,44 @@ internal class XrefResolver
             return (null, resolvedHref, display, xrefSpec?.DeclaringFile);
         }
 
-        public (Error?, ExternalXrefSpec?) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
+        public (Error?, ExternalXrefSpec?, string? href) ResolveXrefSpec(SourceInfo<string> uid, Document referencingFile)
         {
             var (error, xrefSpec) = Resolve(uid, referencingFile);
-            return (error, xrefSpec?.ToExternalXrefSpec());
+            if (xrefSpec == null)
+                return (error, null, null);
+
+            var externalXrefSpec = xrefSpec.ToExternalXrefSpec();
+            var (href, _, fragment) = UrlUtility.SplitUrl(externalXrefSpec.Href);
+            href = RemoveSharingHost(href, _config.HostName);
+            if (UrlUtility.GetLinkType(href) != LinkType.External)
+            {
+                href = PathUtility.Normalize(
+                    UrlUtility.MergeUrl(PathUtility.GetRelativePathToFile(referencingFile.SiteUrl, href), "", fragment));
+            }
+            else
+            {
+                href = UrlUtility.MergeUrl(href, "", fragment);
+            }
+
+            return (error, externalXrefSpec, href);
         }
 
         public XrefMapModel ToXrefMapModel()
         {
-            string? repositoryBranch = null;
-            string? basePath = null;
+            var repositoryBranch = _repository?.Branch;
+            var basePath = _config.BasePath.ValueWithLeadingSlash;
+
             var references = _internalXrefMap.Value.Values
                 .Select(xref =>
                 {
-                    var xrefSpec = xref.ToExternalXrefSpec();
-                    if (repositoryBranch is null)
-                    {
-                        repositoryBranch = _repository?.Branch;
-                    }
-                    if (basePath is null)
-                    {
-                        basePath = _config.BasePath.ValueWithLeadingSlash;
-                    }
-
                     // DHS appends branch infomation from cookie cache to URL, which is wrong for UID resolved URL
                     // output xref map with URL appending "?branch=master" for master branch
-                    var (_, _, fragment) = UrlUtility.SplitUrl(xref.Href);
-                    var path = $"https://{_xrefHostName}{xref.DeclaringFile.SiteUrl}";
+                    var (href, _, fragment) = UrlUtility.SplitUrl(xref.Href);
+                    var path = $"https://{_xrefHostName}{href}";
                     var query = repositoryBranch == "master" ? "?branch=master" : "";
-                    xrefSpec.Href = UrlUtility.MergeUrl(path, query, fragment);
+                    href = UrlUtility.MergeUrl(path, query, fragment);
```

Comment:
`MergeUrl` will split target URL for you:
```suggestion
                    var href = UrlUtility.MergeUrl($"https://{_xrefHostName}{xref.Href}", query, fragment);
```